external/js/cothority/src/byzcoin/v2/darcsBS.ts

```diff
@@ -0,0 +1,268 @@
+        sbs.subscribe(this);
+    }
+}
+
+/**
+ * A DarcBS class represents a darc on byzcoin. It has methods to modify the darc by
+ * adding and removing rules, as well as to change the description.
+ */
+export class DarcBS extends BehaviorSubject<DarcInst> {
+
+    /**
+     * Retrieves a DarcBS from ByzCoin given an InstanceID.
+     *
+     * @param bc of an initialized ByzCoinRPC instance
+     * @param darcID a fixed InstanceID representing the baseID of the darc to retrieve
+     * @return a DarcBS or undefined if something went wrong (no Darc at that ID)
+     */
+    static async retrieve(bc: ByzCoinRPC, darcID: InstanceID):
+        Promise<DarcBS> {
+        Log.lvl3("getting darcBS");
+        const instObs = (await bc.instanceObservable(darcID)).pipe(
+            map((proof) => (proof && proof.value && proof.value.length > 0) ?
+                new DarcInst(Instance.fromProof(darcID, proof)) : undefined),
+        );
+        const bsDarc = await ObservableToBS(instObs);
+        if (bsDarc.getValue() === undefined) {
+            throw new Error("this darc doesn't exist");
+        }
+        return new DarcBS(bsDarc);
+    }
+
+    constructor(darc: BehaviorSubject<DarcInst>) {
+        super(darc.getValue());
+        darc.subscribe(this);
+    }
+
+    /**
+     * Creates an instruction in the transaction with either an update of the description and/or an update
+     * of the rules.
+     *
+     * @param tx where the instruction will be appended to
+     * @param updates contains a description update and/or rules to be merged
+     * @param unrestricted if true, will create an unrestricted evolve that allows to create new rules
+     * @return the new DARC as it will appear on ByzCoin if the transaction is accepted
+     */
+    evolve(tx: TransactionBuilder, updates: IDarcAttr, unrestricted = false): Darc {
+        return DarcInst.evolve(tx, this.getValue(), updates, unrestricted);
+    }
+
+    /**
+     * Sets the description of the DARC.
+     *
+     * @param tx where the instruction will be appended to
+     * @param description of the new DARC
+     * @return the new DARC as it will appear on ByzCoin if the transaction is accepted
+     */
+    setDescription(tx: TransactionBuilder, description: Buffer): Darc {
+        return this.evolve(tx, {description});
+    }
+
+    /**
+     * Sets the sign and evolve rules of the darc.
+     *
+     * @param tx where the instruction will be appended to
+     * @param idSign the IIdentity or the InstanceID of the DARC that is delegated
+     * @param idEvolve per default the same as idSign, but can be given as a different argument
+     * @return the new DARC as it will appear on ByzCoin if the transaction is accepted
+     */
+    setSignEvolve(tx: TransactionBuilder, idSign: IIdentity | InstanceID, idEvolve = idSign): Darc {
+        const rules = this.getValue().rules.clone();
+        rules.setRule(Darc.ruleSign, toIId(idSign));
+        if (idEvolve) {
```

Comment:
is always != undefined

Reply:
no, you can pass `undefined` in the method.

Reply:
if you pass `undefined`, it will get assigned `idSign`, which typewise can't be `undefined`

Reply:
OK - need to test - I was sure it works...

Reply:
Makes sense - tests you are right.

Even though I'd like to see `ts` behave so that I can give `undefined` there ;)

Reply:
Haa, that's ECMA standard (and classic JS way to do it); if you want to retain the two behavior, you can spit the function into `setSign` and `setEvolve`

Reply:
I'm trying to take a `Map`, or an `Object` now, with `action: expression` syntax, so that it's easy callable.

The problem with `setSign` and `setEvolve` is that the darc-version would change twice, and then it gets difficult keeping track of the intermediate version...

Currently the system just supposes that you only update the darc once per transaction. If you update it twice, it will be refused, as the versions will not match.

Reply:
Hum, I'm thinking of removing it and pushing people to use `evolve` directly then, but that might be a bit cumbersome.

The easy way would be to have `idEvolve?` as argument then you can do the undefined check yourself, that seems easier.